thinc/util.py

```diff
@@ -203,12 +203,21 @@ def copy_array(dst: ArrayXd, src: ArrayXd) -> None:  # pragma: no cover
         numpy.copyto(dst, src)  # type: ignore
 
 
-def to_categorical(Y: IntsXd, n_classes: Optional[int] = None) -> FloatsXd:
+def to_categorical(
+    Y: IntsXd,
+    n_classes: Optional[int] = None,
+    label_smoothing: Optional[float] = 0.0,
+) -> FloatsXd:
     xp = get_array_module(Y)
     if n_classes is None:
         n_classes = int(numpy.max(Y) + 1)  # type: ignore
     # Unfortunately, cupy does not support put_along_axis.
-    return xp.eye(n_classes, dtype="float32")[Y]
+    one_hot = xp.eye(n_classes, dtype="float32")[Y]
+    if label_smoothing > 0.0:
+        z = label_smoothing / (n_classes - 1)
+        xp.maximum(0, one_hot - label_smoothing - z, out=one_hot)
```

Comment:
I like the implementation that I suggested more. I had to read this implementation several times to understand how it works. The other implementation is just: every probability is `label_smoothing / (n_classes - 1)`, except the diagonal, which is `1 - label_smoothing`.

This also does many more allocations and operations.

The other implementation:

* Allocate + fill`label_distr`
* Fill diagonal
* Allocate + fill `label_dist[Y]`

This implementation:

* Allocate + fill identity matrix
* Allocate + fill `one_hot`
* Allocate and compute `one_hot - label_smoothing`
* Allocate and compute `one_hot - label_smoothing - z`
* Compute maximum 0 and previous step of `one_hot`
* Add `z` to `one_hot`

Reply:
We should probably also check that `label_smoothing >= 0` and `label_smoothing < 1`.